src/Rules/Classes/EnumSanityRule.php

```diff
@@ -122,6 +124,86 @@ public function processNode(Node $node, Scope $scope): array
 			}
 		}
 
+		$enumCases = [];
+		foreach ($node->stmts as $stmt) {
+			if (!$stmt instanceof Node\Stmt\EnumCase) {
+				continue;
+			}
+			$caseName = $stmt->name->name;
+
+			if (($stmt->expr instanceof Node\Scalar\LNumber || $stmt->expr instanceof Node\Scalar\String_)) {
+				if ($node->scalarType === null) {
+					$errors[] = RuleErrorBuilder::message(sprintf(
+						'Enum %s is not backed, but %s::%s has value %s',
+						$node->namespacedName->toString(),
+						$node->namespacedName->toString(),
+						$caseName,
+						$stmt->expr->value,
+					))
+						->identifier('enum.nonBackedEnumHasValue')
+						->line($stmt->getLine())
+						->nonIgnorable()
+						->build();
+				} else {
+					$caseValue = $stmt->expr->value;
+
+					if (!isset($enumCases[$caseValue])) {
+						$enumCases[$caseValue] = [];
+					}
+
+					$enumCases[$caseValue][] = $caseName;
+				}
+			}
+
+			if ($node->scalarType === null) {
+				continue;
+			}
+
+			if ($node->scalarType->name === 'int' && !($stmt->expr instanceof Node\Scalar\LNumber)) {
+				$errors[] = RuleErrorBuilder::message(sprintf(
+					"Enum case %s::%s doesn't match the '%s' type",
+					$node->namespacedName->toString(),
+					$caseName,
+					'int',
+				))
+					->identifier('enum.caseTypeMismatch')
+					->line($stmt->getLine())
+					->nonIgnorable()
+					->build();
+			}
+
+			// phpcs:ignore SlevomatCodingStandard.ControlStructures.EarlyExit.EarlyExitNotUsed
+			if ($node->scalarType->name === 'string' && !($stmt->expr instanceof Node\Scalar\String_)) {
```

Comment:
Inspired by: https://github.com/phpstan/phpstan-deprecation-rules/blob/9d366debb01a8097106174e46c9aa138d3bf1c01/src/Rules/Deprecations/FetchingDeprecatedConstRule.php#L29

Let me know if it should be rewritten according to the linter rule